Src/FluentAssertions/Data/DataRowAssertions.cs

```diff
@@ -16,13 +16,37 @@ namespace FluentAssertions.Data
     /// used to assert equivalency and the presence of columns.
     /// </summary>
     [DebuggerNonUserCode]
-    public class DataRowAssertions : ReferenceTypeAssertions<DataRow, DataRowAssertions>
+    public class DataRowAssertions<TDataRow> : ReferenceTypeAssertions<TDataRow, DataRowAssertions<TDataRow>>
+        where TDataRow : DataRow
     {
-        public DataRowAssertions(DataRow dataRow)
+        public DataRowAssertions(TDataRow dataRow)
             : base(dataRow)
         {
         }
 
+        /// <summary>
+        /// Returns a <see cref="DataRowAssertions{TDerivedDataRow}"/> object that can be used to assert the
+        /// current <typeparamref name="TDataRow"/> with a compile-time type of <typeparamref name="TDerivedDataRow"/>.
+        /// </summary>
+        /// <typeparam name="TDerivedDataRow">A more derived <see cref="DataRow"/> type to use as the compile-time type for assertions.</typeparam>
+        /// <param name="because">
+        /// A formatted phrase as is supported by <see cref="string.Format(string,object[])" /> explaining why the assertion
+        /// is needed. If the phrase does not start with the word <i>because</i>, it is prepended automatically.
+        /// </param>
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <paramref name="because"/>.
+        /// </param>
+        public DataRowAssertions<TDerivedDataRow> As<TDerivedDataRow>(string because = "", params object[] becauseArgs)
```

Comment:
I agree with the gut feeling, but this is where the problems begin with generic types and `Should<T>` only supporting one set of constraints. For typed data sets, tables and rows, we would need:
```
myDataSet.Should().BeEquivalentTo(...)
           ^-- return `DataSetAssertions<MyDataSetType>

myDataTable.Should().BeEquivalentTo(...)
              ^-- return `DataTableAssertions<MyDataTableType>`

myDataRow.Should().BeEquivalentTo(...)
           ^-- return `DataRowAssertions<MyDataRowType>`
```
This requires there to be three different `Should<T>` overloads that differ only in their `where` clause.

Hmm, maybe that would be possible to do if they were defined in separate classes? Investigating...

Reply:
My idea is that if one wants to assertions on `MyDataRow` instead of just `DataRow`, they just add this to _their_ unit test project.

```cs
public static class ShouldExtensions
{
    public static DataRowAssertions<MyDataRow> Should(this MyDataRow value) => 
        new DataRowAssertions<MyDataRow>(value);
}
```

Would that work?

Reply:
My thinking is that if _that_ would work, it should also work for FluentAssertions itself to just have more than one static class providing extension methods, so that `Should<T>` _can_ have different overloads that differ only in their constraints, just not in the same class. I'm about to test this theory. :-)

Reply:
This seems to work: 7e22676a